app/scripts/models/flow.js

```diff
@@ -0,0 +1,130 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * A model to represent the somewhat nebulous concept of an
+ * "event flow". A single flow could be a user connecting a
+ * new device to their Sync account, for instance.
+ *
+ * Each flow has two properties.
+ *
+ *   * flowId: 32-byte identifier hex-encoded as a string.
+ *   * flowBeginTime: server-generated timestamp.
+ *
+ * When a new flow is begun, it should result in an activity
+ * event being logged by the server. To that end, we push a
+ * a `flow.begin` onto the metrics event array.
+ *
+ * There is some mess surrounding when to begin a new flow,
+ * which necessitates equivalent mess in this model. It runs
+ * something like this:
+ *
+ *   1. A user arriving at /signin or /signup constitutes the
+ *      beginning of a flow, so a corresponding `flow.begin`
+ *      event must be logged.
+ *
+ *   2. A user refreshing the page on /signin or /signup does
+ *      not constitute a new flow, so events must not be logged
+ *      in those cases. This is achieved by saving the flowId
+ *      in local storage.
+ *
+ *   3. If a user signs in, then signs out, then signs in again,
+ *      the second sign-in does count as a new flow, so a second
+ *      `flow.begin` event must be logged. This is achieved by
+ *      clearing the model and local storage when a user is signed
+ *      out. Also note that in this case, since the navigation may
+ *      have occurred entirely on the client, we might not have a
+ *      server-generated flowBeginTime and must estimate the time
+ *      instead.
+ *
+ *   4. If a user signs up on device A then confirms their email
+ *      address on device B, the flow id must follow them onto
+ *      the new device so that it can be propagated in requests
+ *      to the auth server. This is achieved by also setting flowId
+ *      as a property of the user model, which is serialised into
+ *      the resume token.
+ */
+
+define(function (require, exports, module) {
+  'use strict';
+
+  var _ = require('underscore');
+  var Backbone = require('backbone');
+  var Cocktail = require('cocktail');
+  var ResumeTokenMixin = require('models/mixins/resume-token');
+  var SearchParamMixin = require('models/mixins/search-param');
+  var Storage = require('lib/storage');
+  var uuid = require('uuid');
+  var vat = require('lib/vat');
+
+  var STRATEGIES = [
+    populateFromResumeToken,
```

Comment:
> used for all time,

This is really key and highlights something that was woefully under-specified when I pitched this feature.  We've made great progress on specifying when a flow should *start*, but we haven't dug into when a flow should *stop*.

Zooming out, the high-level motivation for this feature is to tie together the individual events along the path to connecting one or more devices, so that we can generate two kinds of graph:

* Show the percentage dropoff at each stage of the connection funnel
* Investigate the time between each stage of the connection funnel

If it's not going to be useful to compare two events along either of those axes, we don't need to go out of our way to give them the same flowid.  Some things that might mean events are not related in this way:

* Closing the tab, or navigating away from the tab
* When too much time has passed, e.g. 5 or 10 minutes with no activity
* Maybe certain success states in the page itself?  (e.g. if you're successfully viewing the /settings page, you've finished whatever flow you were doing)